File: csharp/src/Training.Workshop/M8/S04ShadowLimits/Step3/ShadowBooking.cs
namespace Training.Workshop.M8.S04ShadowLimits.Step3;

/// <summary>
/// Krok 3: cień woła tylko czysty BookingPlanner - bez nagrywarek i bez ryzyka, że ktoś
/// podepnie prawdziwy adapter. Podwójne wykonanie dotyczy wyłącznie obliczenia.
/// </summary>
public sealed class ShadowBooking
{
    private readonly Infrastructure _infra;
    private readonly LegacyBookingFlow _legacy;
    private readonly BookingPlanner _candidate = new();
    private readonly List<string> _divergences = [];

    public ShadowBooking(Infrastructure infra)
    {
        _infra = infra;
        _legacy = new LegacyBookingFlow(infra);
    }

    public string Book(BookingRequest request)
    {
        int before = _infra.Log().Count;
        string result = _legacy.Book(request);
        IReadOnlyList<string> legacyEffects = _infra.Log().Skip(before).ToList();
        try
        {
            BookingPlan plan = _candidate.Plan(request);
            Compare("wynik", [result], [plan.Result]);
            IReadOnlyList<string> planned = plan.Effects.Select(effect => effect.Describe()).ToList();
            Compare("efekty", legacyEffects, planned);
        }
        catch (Exception failure)
        {
            _divergences.Add("kandydat: " + failure.Message);
        }
        return result;
    }

    private void Compare(string what, IReadOnlyList<string> legacyValues, IReadOnlyList<string> candidateValues)
    {
        for (int i = 0; i < Math.Max(legacyValues.Count, candidateValues.Count); i++)
        {
            string legacyValue = i < legacyValues.Count ? legacyValues[i] : "-";
            string candidateValue = i < candidateValues.Count ? candidateValues[i] : "-";
            if (!legacyValue.Equals(candidateValue))
            {
                _divergences.Add(what + " legacy: " + legacyValue + " | kandydat: " + candidateValue);
            }
        }
    }

    public IReadOnlyList<string> Divergences()
    {
        return _divergences.ToList();
    }
}

File: typescript/src/workshop/m8/s04_shadowlimits/step3/ShadowBooking.ts
import type { BookingRequest } from '../BookingRequest.js';
import type { Infrastructure } from '../Infrastructure.js';
import { BookingPlanner } from './BookingPlanner.js';
import { LegacyBookingFlow } from './LegacyBookingFlow.js';

/**
 * Krok 3: cień woła tylko czysty BookingPlanner - bez nagrywarek i bez ryzyka, że ktoś
 * podepnie prawdziwy adapter. Podwójne wykonanie dotyczy wyłącznie obliczenia.
 */
export class ShadowBooking {
  private readonly legacy: LegacyBookingFlow;
  private readonly candidate = new BookingPlanner();
  private readonly divergenceList: string[] = [];

  constructor(private readonly infra: Infrastructure) {
    this.legacy = new LegacyBookingFlow(infra);
  }

  book(request: BookingRequest): string {
    const before = this.infra.log().length;
    const result = this.legacy.book(request);
    const legacyEffects = this.infra.log().slice(before);
    try {
      const plan = this.candidate.plan(request);
      this.compare('wynik', [result], [plan.result]);
      const planned = plan.effects.map((effect) => effect.describe());
      this.compare('efekty', legacyEffects, planned);
    } catch (failure) {
      this.divergenceList.push('kandydat: ' + (failure as Error).message);
    }
    return result;
  }

  private compare(what: string, legacyValues: readonly string[], candidateValues: readonly string[]): void {
    for (let i = 0; i < Math.max(legacyValues.length, candidateValues.length); i++) {
      const legacyValue = i < legacyValues.length ? legacyValues[i]! : '-';
      const candidateValue = i < candidateValues.length ? candidateValues[i]! : '-';
      if (legacyValue !== candidateValue) {
        this.divergenceList.push(what + ' legacy: ' + legacyValue + ' | kandydat: ' + candidateValue);
      }
    }
  }

  divergences(): readonly string[] {
    return Object.freeze([...this.divergenceList]);
  }
}
